content/en/docs/plugins/developers_guide.md

```diff
@@ -0,0 +1,1081 @@
+	//
+	// Empty configs are allowed, in which case the default is
+	// used.
+	if cfg != "" && cfg != "{}" {
+		var obj map[string]uint64
+		err := json.Unmarshal([]byte(cfg), &obj)
+		if err != nil {
+			return err
+		}
+		if _, ok := obj["jitter"]; ok {
+			jitter = obj["jitter"]
+		}
+	}
+
+	m.jitter = jitter
+	m.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	return nil
+}
+
+func (m *MyPlugin) Destroy() {
+	log.Printf("[%s] Destroy\n", PluginName)
+}
+```
+
+#### Opening/Closing a Stream of Events
+
+A plugin instance is created when the plugin event stream is opened, which can happen more than once during the plugin lifecycle. Source plugins are required to define an `Open()` method that creates a returns a new plugin instance. This is where the framework passes in the user-defined open parameters string.
+
+The plugin instance type returned form `Open()` can define an optional `Close()` method bundling any additional deinitialization logic to run while closing the event stream.
+
+```go
+func (m *MyPlugin) Open(prms string) (source.Instance, error) {
+	log.Printf("[%s] Open, params=%s\n", PluginName, prms)
+
+	// The format of params is a json object with two params:
+	// - "start", which denotes the initial value of sample
+	// - "maxEvents": which denotes the number of events to return before EOF.
+	// Example:
+	// {"start": 1, "maxEvents": 1000}
+	var obj map[string]uint64
+	err := json.Unmarshal([]byte(prms), &obj)
+	if err != nil {
+		return nil, fmt.Errorf("params %s could not be parsed: %v", prms, err)
+	}
+	if _, ok := obj["start"]; !ok {
+		return nil, fmt.Errorf("params %s did not contain start property", prms)
+	}
+
+	if _, ok := obj["maxEvents"]; !ok {
+		return nil, fmt.Errorf("params %s did not contain maxEvents property", prms)
+	}
+
+	return &MyInstance{
+		initParams: prms,
+		maxEvents:  obj["maxEvents"],
+		counter:    0,
+		sample:     obj["start"],
+	}, nil
+}
+
+func (m *MyInstance) Close() {
+	log.Printf("[%s] Close\n", PluginName)
+}
+```
+
+#### Returning Events
+
+New events are generated in batch by the `NextBatch` function. The function is mandatory for source plugins and must be defined as a method of the plugin instance struct type. The `pState` argument is the plugin struct type initialized in `Init()`, passed in by the framework for ease of access. The plugin state is passed as an instance of the `sdk.PluginState` interface, so a manual cast is required to access the internal state variables defined in the struct type.
+
+The `evts` parameter is a sdk-managed batch of events to be used for creating new events. For that, the SDK uses a slab allocator and reuses the same event batch at every iteration to improve performance. The lenght of the `evts` list represents the maximum size of each event batch.
+Each element of the batch is an instance of `sdk.EventWriter` that provides handy methods to write the event info and data. Event data can be written with the Go `io.Writer` interface.
```

Comment:
I think it might be useful to document and describe the EventWriter(s) interfaces (and maybe EventReader too) as a section "Interacting with events". You could talk about the slab allocator and how it's very memory efficient, as long as someone uses the interfaces to access events.

Reply:
The new section looks great, but I would move it into the sdk walkthrough instead of the dummy plugin walkthrough.

Reply:
thanks for moving it!